---
title: "Risk in relatives, heritability, SNP-based heritability and genetic correlations in psychiatric disorders: a review"
author: "Bart Baselmans, Wouter van Rheenen & Naomi Wray"

#output: pdf_document
output:
   html_document:
     toc: true
     toc_depth: 2
     toc_float: true
     number_sections: true
     smooth_scroll : true
     code_folding: "hide"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction


This supplementary note provides more details on the content discussed in the paper *Risk in relatives, heritability, SNP-based heritability and genetic correlations in psychiatric disorders: a review*, published in *Biological Psychiatry*. Click on "Code"" to expand to see the R code.

```{r echo=FALSE, message=FALSE, warning=FALSE}
# You need these libraries for html
library(rmarkdown)    # install.packages("rmarkdown") 
library(data.table)
library(knitr)
#To knit to PDF you need to install: MaxTeX (http://www.tug.org/mactex/) or TeX (https://www.tug.org/texlive/)

```

# Lifetime risk and increased risk in developing psychiatric disorders
In Figure 1 of the paper we show estimates of the population lifetime risk ($K$) and the risk in first degree relatives of those with one affected parent ($K_{1}$), and the risk ratio ($RR_{1} = \lambda_{1}=K_{1}/K$) for common psychiatric disorders. Although $K$ and the risk in relatives of different types ($K_R$) are measurable, often only the heritability estimated from one or more types of relatives are reported. Heritability is defiend as the proportion of variance in liability attributed to additive genetic factors. Here, we present the $K$ and heritabilities for the major psychiatric disorders (see Supplementary Table 1 for references). The estimates using twin data are usually higher than those estimated from family data (different types of relatives) although the latter are often based on larger sample sizes. Here we used approximate round $h^2$ numbers (average $h^2_{twin}$ and $h^2_{family}$) in subsequent calculations. We also provide SNP-based heritability ($h^2_{SNP}$)  estimates, where SNP-based heritability, by definition is smaller than the estimates based on family records, as it tracks only the proportion of variance attributable to additives genetic values tagged by common SNPs (or other measurable common DNA variants):

Phenotype | $h^2_{twin}$ (in %) | $h^2_{family}$ (in %) |$h^2_{round}$ | $h^2_{SNP}$ in % | Lifetime risk ($K$)  |  
|:----------|:----------|:----------|:----------|:----------|:----------|
**Schizophrenia** | 81 | 64 |70 |22 | 0.01 |
**Bipolar Disorder** | 75 | 59 | 65 |18 | 0.01 |
**Major Depressive Disorder** | 37 | 32 |35| 8.5 | 0.15
**ADHD** | 75 | - |75| 22 | 0.05 |
**Anorexia Nervosa** |56 | 43|50| 20 | 0.01
**Autism Spectrum Syndrome** | 80 | 85 |80| 11 | 0.01

## Liability Threshold model

When many factors contribute to the risk of disease it can be helpful to consider a model of disease where there is a latent distribution of liability to disease representing the dichotomous Case/Control status.(e.g. [Falconer, 1965](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1469-1809.1965.tb00500.x) or [Reich 1972](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1469-1809.1972.tb00767.x)). Since this latent liability comprises many genetic and other risks it is reasonable to assume that the liability distribution is approximately normally distributed (since many things added together will make a bell-shaped distribution in a population sample) and that those affected by disease have the combination of genetic and other factors that it places them in the top end of the liability distribution. 


```{r, liability_model, fig.height=4, fig.width=10, fig.align="center", echo=T}

h2x = 0.7
h2y = 0.35
Kx = 0.01
Ky = 0.15
disorder1 = "Schizophrenia"
disorder2 = "Major Depressive Disorder"

plot_liability_model <- function(h2x,h2y,Kx,Ky, disorder1, disorder2){

Tx = -qnorm(Kx, 0,1)
Ty = -qnorm(Ky, 0,1)

Ks <- c(Kx,Ky)  #Lifetime risk
Heritability <- c(h2x,h2y)
Threshold <-c(Tx,Ty)
disorder <-c(disorder1,disorder2)

layout(matrix(c(1,2),1, 2, byrow = TRUE))
for(j in 1:2){

h2 <- c(as.expression(bquote(italic(h)^2~'='~.(Heritability[j]))),
                  as.expression(bquote(italic(h)^2~'='~.(Heritability[j]))))
K_pop <- c(as.expression(bquote(italic(K)~'='~.(Ks[j]))),
                    as.expression(bquote(italic(K)~'='~.(Ks[j]))))

  T0 = Threshold[j]
  z = dnorm(T0) 
  i = z / Ks[j]
  mean = -(T0) ;sd=1
  lb=0; ub=4
  
  x <- seq(T0-6,T0+3,length=1000)*sd + mean
  hx <- dnorm(x,mean,sd)
  
  plot(x, hx, type="n", xlab="", ylab="",
       main="", axes=FALSE)
  
  i <- x >= lb & x <= ub
  l <- x <= lb & x <= ub
  lines(x, hx)
  polygon(c(lb,x[i],ub), c(0,hx[i],0), col="darkblue")
  polygon(c(-5,x[l],lb), c(0,hx[l],0), col="lightgrey")
  
  
  #axis(1, at=seq(-5, 2, 1), pos=0, cex.axis =1)
  abline(h=0)
  abline(v=0,h=-2)
  
  mtext(K_pop[j], side = 3, line = -2.1,at = 2,font = 2, cex = 0.8, adj=1)
  mtext(disorder[j], side = 1, line = 1.5,at = 0,font = 2, cex = 0.8, adj=1)
  }
}

plot_liability_model(h2x,h2y,Kx,Ky, disorder1, disorder2)

```

Hence, those whose  liability to disease is above the threshold are affected, and so this model is sometimes called the liability threshold model. The threshold can be calculated using the lifetime risk ($K$) parameter as $K$ represents the proportion of individuals in the general population have the disorder of interest. The threshold (or boundary) that determines the area of $K$ (indicated in blue) in a normal distribution is given by:
$$
T=\Phi^{-1}(1-K)
$$

Conversely, the life-time risk ($K$) is derived from the threshold parameter
$$
K = 1-\Phi(T)
$$
where $\Phi^{-1}(x)$ is the inverse of the cummulative standard normal distribution function. 


An example with R code using lifetime risk $K = 0.15$:
```{r}
K = 0.15
Threshold_prevalence <- function(K){
Tx = -qnorm(K,0,1)
Kx = 1-pnorm(Tx,0,1)

cat("Output Parameters:\n")
cat("----------------------------\n")
cat("Tx: ", round(Tx,2),"\n")
cat("Kx: ", round(Kx,2),"\n")
cat("----------------------------\n")    
}
Threshold_prevalence(K)
```

## Risk in relatives and heritability of liability

The lifetime risk in relatives ($K_{R}$) from population samples. It is logical to assume that the threshold in liability associated with disease has the same value. As a result, the liability distribution in first degree members must be shifted (in the direction of increased liability) compared to the population in consistent with the observed higher risk of $K_{R}$.
The shift in mean distribution is $a_Rih^2$, where $a_R$ is the genetic relationship between the relatives (i.e., 0.5 for 1st degree relatives), $i$ is the mean phenotypic liability of those with disease, which is calculated as $z/K$, where $z$ is the height of the standard normal curve at threshold $T$. Hence, the shift in mean liability between population and 1st degree relatives is higher for diseass with higher heritability and for diseases that are less common.  

The difference in mean liabilities between population and 1st degree relatives, can be shown to be equivalent to the difference in thresholds when the liability distribution of relatives is scaled back to a N(0,1) distribution. Hence,$T_1=\Phi^{-1}(1-K_1)$ and according to [Falconer (1965)](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1469-1809.1965.tb00500.x) is simply $T_R - T = a_Rih^2$, so that heritability $h^2$ can be estimated as
$h^2 = \frac{T - T_R}{a_Ri}$, i.e. everything on the right hand side of the equation is derived from measurable statistics.

However, [Reich et al.,(1972)](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1469-1809.1972.tb00767.x)  showed that while it is OK to assume that the distribution of disease liability in relatives of those affected is approximately normal, it should be recognised that the variance in liability is slightly reduced in the relatives as a result of ascertaining the proband as been affected. They showed that the variance in liability is reduced by a factor of $a_{R}^2h^4i(i-T)$ 
so that:


$$
T_{R} = \frac{T -a_Rih^2}{\sqrt{1-a_{R}^2h^4i(i-T)}}
$$
Since heritability is in the numerator and denominator, making $h^2$ the subject of the equation requires solving of a quadratic equation to give:
$$
h^2 = \frac{T - T_R \sqrt{1 -(1-\frac{T}{i})(T^2-T^2_R)}}{a_R(i+(i-T)T^2_R)}
$$
Although, this equation looks complicated, in fact everything on the right hand side can be calculated from two observations $K$ and $K_R$, and this equation holds as long as $a_R$ < 1 or if ascertainment is based on the disease status of one proband. A special case is when ascertainment is based on both parents being affected, hence the $K$ and $K_{2PAR}$ is estimated, with risk ratio $RR_{2PAR} = \frac{K_{2PAR}}{K}$
In this case it can be shown [Wray & Gottesman (2011)](https://www.frontiersin.org/articles/10.3389/fgene.2012.00118/full) that

$$
T_{2PAR} = \frac{T-ih^2}{\sqrt{1-0.5h^4i(i-T)}}
$$
and solving this quadratic equation gives:

$$
h^2 = 2T-\sqrt{2} T_{2PAR} \frac{\sqrt{2 -(T^2 - T_{2PAR})(1-\frac{T}{i})}}{2i+(i-T)T_{2PAR}}
$$
\
\
To visualize the increase liability for relatives having one or two affected parents consider examples of schizophrenia with $K=0.01$,and $h^2=0.7$ or major depressive disorder with $K=0.15$, $h^2=0.35$  a sample of 100 individuals drawn from the general population, and a sample of 100 children who have one affected parent.   


```{r, liability_model2, fig.height=7, fig.width=8, fig.align="center", echo=T}
h2x = c(0.7,0.35)
Kx = c(0.01,0.15)
a = 0.5
disorder = c("Schizophrenia","Major Depressive Disorder")


#################################################################################################
liability <- c()
liability_function <- function(h2x, Kx, a, disorder){

  Tx    = -qnorm(Kx, 0,1)
  z    = dnorm(Tx)
  i    = z/Kx
#One affected parent
  Tx1   = (Tx - a * i * h2x) / (sqrt(1 - a * a * h2x * h2x * i * (i-Tx)))
  Kx1   = 1 - pnorm(Tx1)
  RRx1   = Kx1 / Kx
#Two affected parents  
  Tx2   = (Tx -i*h2x) / sqrt(1-(0.5*h2x*h2x*i)*(i-Tx))
  Kx2   = 1 - pnorm(Tx2)
  RRx2   = Kx2 / Kx
  
  x <- as.data.frame(c(disorder,Tx,Tx1, Tx2, Kx, Kx1, Kx2, NA,RRx1, RRx2 ))
  return(x) 
}  

#Run function for number of included disorders
for(i in 1:length(disorder)){
liability[i] <- liability_function(h2x[i], Kx[i], a, disorder[i])
}

#convert list to dataframe
df <- data.frame(matrix(unlist(liability), nrow=length(liability), byrow=T), stringsAsFactors = F)

#Threshold extraction
Threshold <- df[2:4]
Threshold_row <-cbind(Threshold[1,], Threshold[2,])
Threshold_row  <- as.numeric(Threshold_row)

# Prevalence extraction
K_prevalence <- df[5:7]
K_prevalence_row <- as.numeric(cbind(K_prevalence[1,], K_prevalence[2,]))
K100_prevalence_row <- round((K_prevalence_row*100),0)

#Make plot
layout(matrix(c(1,2,3,7,8,9,4,5,6,10,11,12), 3, 4, byrow = F))
par(mar=c(5.1, 4.1, 4.1, 2.2))
for(j in 1:6){

#Normal distribution Threshold  
  Ks <- as.matrix(sapply((K100_prevalence_row/100), as.numeric))  
  K_print <- c(as.expression(bquote(italic(K)[SCZ]~'='~.(Ks[j]))),
               as.expression(bquote(italic(K)[1][PAR][-SCZ]~'='~.(Ks[j]))),
               as.expression(bquote(italic(K)[2][PAR][-SCZ]~'='~.(Ks[j]))),
               as.expression(bquote(italic(K)[MDD]~'='~.(Ks[j]))),
               as.expression(bquote(italic(K)[1][PAR][-MDD]~'='~.(Ks[j]))),
               as.expression(bquote(italic(K)[2][PAR][-MDD]~'='~.(Ks[j]))))
  
  heritability1 <- c(paste0("= ",h2x[1]))
  Phenotype1<- c("","",as.expression((bquote(bold(bold(.(disorder[1])~'('*bold(italic(h)^2)~.(heritability1)*')')))),"",""))
  heritability2 <- c(paste0("= ",h2x[2]))
    Phenotype2 <- c("","","","","",as.expression((bquote(bold(bold(.(disorder[2])~'('*bold(italic(h)^2)~.(heritability2)*')')))),"",""))
  
  Txp = Threshold_row[j]
  z = dnorm(Txp) 
  i = z / Ks[j]
  mean = -(Txp) ;sd=1
  lb=0; ub=4
  
  x <- seq(Txp-6,Txp+3,length=1000)*sd + mean
  hx <- dnorm(x,mean,sd)
  
  plot(x, hx, type="n", xlab="", ylab="",
       main="", axes=FALSE)
  i <- x >= lb & x <= ub
  l <- x <= lb & x <= ub
  lines(x, hx)
  polygon(c(lb,x[i],ub), c(0,hx[i],0), col="darkblue")
  polygon(c(-5,x[l],lb), c(0,hx[l],0), col="grey") 
  
  axis(1, at=seq(-5, 2, 1), pos=0, cex.axis =1)
  abline(h=0)
  abline(v=0,h=-2)
  
  mtext(K_print[j], side = 3, line = -1.1,at = 10,font = 2, cex = 0.8, adj=1)
  mtext(Phenotype1[j], side = 3, line = -12,at = -2.5,font = 2, cex = 0.7, adj = 0)
  mtext(Phenotype2[j], side = 3, line = -12,at = -2.1,font = 2, cex = 0.7, adj = 0)
  
}

############################
##########################

#Square indicating number of affected individuals
result <- matrix(nrow = 6, ncol = 100)
#layout(matrix(c(1,2,3,4,5,6), 3, 2, byrow = F))
for(j in 1:6){
  result[j,] <- sample(as.matrix(c(rep("b",K100_prevalence_row[j]),rep("a",100-K100_prevalence_row[j]))))
  
  
  #sum(result[1,]=="a")
  # input parameters - nr * nc should equal length(x)
  cols <- c("grey", "darkblue")
  nr <- 10
  nc <- 10
  
  # create data.frame of positions and colors
  m <- matrix(cols[factor(result[j,])], nr, nc)
  DF <- data.frame(row = c(row(m)), col = c(col(m)[, nc:1]), value = c(m), gender =  sample(c(rep("male",50),rep("female",50))),
                   stringsAsFactors = FALSE)
  
  title_plot <- c("Sample drawn from \n the population", "Sample drawn from \n those with one affected parent", "Sample drawn from \n those with two affected parents","Sample drawn from \n the population", "Sample drawn from \n those with one affected parent", "Sample drawn from \n those with two affected parents")
  
  plot(col ~ row, DF, col = DF$value, pch = 15, cex = 1, asp = 1,
       xlim = c(0, nr), ylim = c(0, nc),
       axes = FALSE, xlab = "", ylab = "")
  title(xlab=title_plot[j], line=0.2, cex.lab=1)
}

```

For the major psychiatric disorders the increased risk ratio are:

```{r, results= 'asis', echo = T}
h2x = c(0.70,0.65,0.35,0.75,0.5,0.8)
Kx = c(0.01,0.01,0.15,0.05,0.01,0.01)
a = 0.5
disorder = c("Schizophrenia","Bipolar Disorder", "Major Depressive Disorder", "ADHD", "Anorexia Nervosa", "Autism Spectrum Disorder")
liability <- c()
liability_function <- function(h2x, Kx, a, disorder){
  
  Tx    = -qnorm(Kx, 0,1)
  z    = dnorm(Tx)
  i    = z/Kx
  #One affected parent
  Tx1   = (Tx - a * i * h2x) / (sqrt(1 - a * a * h2x * h2x * i * (i-Tx)))
  Kx1   = 1 - pnorm(Tx1)
  RRx1   = round(Kx1 / Kx,1)
  #Two affected parents  
  Tx2   = (Tx -i*h2x) / sqrt(1-(0.5*h2x*h2x*i)*(i-Tx))
  Kx2   = 1 - pnorm(Tx2)
  RRx2   = round(Kx2 / Kx,0)
  
  x <- as.data.frame(c(disorder,(h2x*100), round(Kx,2), round(Kx1,2), round(Kx2,2),round(RRx1,2), round(RRx2,2)))
  return(x) 
}  
#Run function for number of included disorders

for(i in 1:length(disorder)){
  liability[i] <- liability_function(h2x[i], Kx[i], a, disorder[i])
}
#convert list to dataframe
Threshold_model_df <- data.frame(matrix(unlist(liability), nrow=length(liability), byrow=T), stringsAsFactors = F)
colnames(Threshold_model_df) <- c("Disorder","heritability","K", "K1", "K(2par)", "RR", "RR(2par)")

library(knitr)
kable(Threshold_model_df, caption ="Risk ratio in relatives")

```

*$h^2$ estimates taken from family studies except ADHD ($h^2_{twin}$)*

The R code to calculate $T$, $K$, and $RR$ is provided below

```{r T-K-RR}
h2x=0.7
Kx=0.01
a=0.5
disorder="Schizophrenia"

Liability_threshold <- function(h2x, Kx, a, disorder){
  cat("Input Parameters:\n")
  cat("----------------------------\n")
  cat("disorder:", disorder, "\n")
  cat("h2x: ",h2x,"\n")
  cat("Kx : ",Kx,"\n")
  cat("IBD: ",a, "\n")
  cat("----------------------------\n")  
  Tx    = -qnorm(Kx, 0,1)
  z    = dnorm(Tx)
  i    = z/Kx
  Tx1   = (Tx - a * i * h2x) / (sqrt(1 - a * a * h2x * h2x * i * (i-Tx)))
  Kx1   = 1 - pnorm(Tx1)
  RR   = Kx1 / Kx
  
  Tx2   = (Tx -i*h2x) / sqrt(1-(0.5*h2x*h2x*i)*(i-Tx))
  Kx2   = 1 - pnorm(Tx2)
  RR2   = Kx2 / Kx
  # return(matrix(data = c(Tx,Kx,NA, Tx1, Kx1,RR, Tx2, Kx2,RR2), nrow = 3, ncol = 3 ))
  
  cat("Output Parameters:\n")
  cat("----------------------------\n")
  cat("T       : ",round(Tx,2),"\n")
  cat("T(1)    : ",round(Tx1,2),"\n")
  cat("T(2par) : ",round(Tx2,2), "\n")
  cat("K(1)    : ",round(Kx1,2),"\n")
  cat("K(2par) : ",round(Kx2,2), "\n")
  cat("RR      : ", round(RR,1), "\n")
  cat("RR(2par): ", round(RR2,0), "\n")
  cat("----------------------------\n")    
}

Liability_threshold(h2x,Kx,a, disorder)
```

## Different views of the liability threshold distribution
The liability model assumes that genetic and non-genetic factors contribute additively to liability to disease. This implies a non-lineare relationship between phenotypic and genetic liability with probability of disease. 

Consider that individuals either have disease (probability of disease =1) or do not have disease (probability of disease =0) then when considering the relationship between phenotypic liability and disease, a vertical line will correspond to the threshold $T$ corresponding to the disease given lifetime risk of $K$ that bisects risk of disease (as visualized above). For schizophrenia (red), and major depressive disorder (black) with  $K = 0.01$ and $K=0.15$ the relationship between phenotypic liability and probability of developing the disorder looks like:


```{r, phenotypic_liability, echo=T, fig.width=4, fig.height=4, fig.align="center"}
r2x <- 1
Kxp <- 0.01
r2y <- 1
Kyp <- 0.15
a <- 0.5
disorder1 <- "Schizophrenia"
disorder2 <- "Major Depressive Disorder"

xrange <- seq(-4,+6,len=100)
ProbDisease<- function(K,r2,xrange){
  sapply(xrange,function(x) pnorm( (x-qnorm(1-K))/sqrt(1-r2) ))
}

par(mar=c(5.1, 4.1, 4.1, 2.1))


l3<- c(as.expression((bquote(SCZ~'('*italic(K)~'= 0.01'*')'))))
l4<- c(as.expression((bquote(MDD~'('*italic(K)~'= 0.15'*')'))))
xrange <- seq(-4,+6,len=100)

SCZ<- ProbDisease(K=Kx,r2=r2x,xrange)
MDD <- ProbDisease(K=Ky,r2=r2y,xrange)
  
matplot(xrange,cbind(SCZ,MDD), mgp=c(3,1,0),
          frame.plot = FALSE, type="l",lty=1:2,col=1:2,lwd=3, 
          xlab = "Phenotypic Liability",cex.axis=0.6, cex.lab =0.5,
          ylab = "Probability developing disorder")
          legend("topleft", legend = c(l3,l4) ,
          col =1:2, lty = 1:2, lwd=3, cex = 0.5,bty = "n")



```

Now, if we visualize the relationship between genetic liability and risk/probability of disease you will observe a very non-linear relationship with the gradient of relationship becomes steeper when heritability becomes higher (See difference between major depressive disorder with $h^2 = 0.39$ and schizophrenia with $h^2=0.7$).


```{r, probability, echo=T, fig.width=10, fig.height=4, fig.align="center"}
h2x <- 0.7
Kxp <- 0.01
h2y <- 0.35
Kyp <- 0.15
a <- 0.5
disorder1 <- "Schizophrenia"
disorder2 <- "Major Depressive Disorder"

#Compute relevant parameters
get_Tr_Kr_RR <- function(h2x, Kxp,h2y,Kyp, a, disorder1, disorder2){
  
  Txp    = -qnorm(Kxp, 0,1)
  z    = dnorm(Txp)
  i    = z/Kxp
  Tx1   = (Txp - a * i * h2x) / (sqrt(1 - a * a * h2x * h2x * i * (i-Txp)))
  Kx1   = 1 - pnorm(Tx1)
  RRx1   = Kx1 /Kxp
  
  Tx2   = (Txp -i*h2x) / sqrt(1-(0.5*h2x*h2x*i)*(i-Txp))
  Kx2   = 1 - pnorm(Tx2)
  RRx2   = Kx2 / Kxp
  
  
  Typ    = -qnorm(Kyp, 0,1)
  z    = dnorm(Typ)
  i    = z/Kyp
  Ty1   = (Typ - a * i * h2y) / (sqrt(1 - a * a * h2y * h2y * i * (i-Typ)))
  Ky1   = 1 - pnorm(Ty1)
  RRy1   = Ky1 /Kyp
  
  Ty2   = (Typ -i*h2y) / sqrt(1-(0.5*h2y*h2y*i)*(i-Typ))
  Ky2   = 1 - pnorm(Ty2)
  RRy2   = Ky2 / Kyp

  
  x <- matrix(c(Txp,Tx1, Tx2, Kxp, Kx1, Kx2, NA,RRx1, RRx2), nrow = 3, ncol = 3, byrow = T )
  y <- matrix(c(Typ,Ty1, Ty2, Kyp, Ky1, Ky2, NA,RRy1, RRy2), nrow = 3, ncol = 3, byrow = T )
  xy <- cbind(x,y)       
  return(xy)              
}
Parameters <- get_Tr_Kr_RR(h2x, Kxp,h2y,Kyp, a, disease1, disease2)
Ks <-round(Parameters[2,],2)
KSCZ <- Ks[1:3] 
KMDD <- Ks[4:6] 

xrange <- seq(-4,+6,len=100)
ProbDisease<- function(K,h2,xrange){
  sapply(xrange,function(x) pnorm( (x-qnorm(1-K))/sqrt(1-h2) ))
}

layout(matrix(c(1,2,3), 1, 3, byrow = F))
par(mar=c(5.1, 4.1, 4.1, 2.1))


for(k in 1:3){  
l3<- c(as.expression((bquote('('*italic(K)[SCZ]*')'))),
       as.expression((bquote('('*italic(K)[1][PAR][-SCZ]*')'))),
       as.expression((bquote('('*italic(K)[2][PAR][-SCZ]*')'))))

l4<- c(as.expression((bquote('('*italic(K)[MDD]*')'))),
       as.expression((bquote('('*italic(K)[1][PAR][-MDD]*')'))),
       as.expression((bquote('('*italic(K)[2][PAR][-MDD]*')'))))
       
xrange <- seq(-4,+6,len=100)

SCZ<- ProbDisease(K=KSCZ[k],h2=h2x,xrange)
MDD <- ProbDisease(K=KMDD[k],h2=h2y,xrange)
  
matplot(xrange,cbind(SCZ,MDD), mgp=c(3,1,0),
          frame.plot = FALSE, type="l",lty=1:2,col=1:2,lwd=3, 
          xlab = "Genetic Liability",cex.axis=1, cex.lab =1,
          ylab = "Probability developing disorder")
          legend("topleft", legend = c(l3[k],l4[k]) ,
          col =1:2, lty = 1:2, lwd=3, cex = 0.8,bty = "n")
}

```

# SNP-based heritability

Heritability is the ratio of additive genetic variance $V_A$ divided by the sum of $V_A$ and the residual variance $V_E$:

$$
h^2 = \frac{V_A}{V_A+V_E}
$$

Therefore, the genetic factors unique to an individual but which impact of disease status for that individual (i.e., de novo mutations) would be partitioned into the residual (difference between total and genetic) variance ($V_E$). Estimates of $h^2$ made from family record contain contributions from both rare and common genetic variants, in contrast $h^2_{SNP}$ contains only the proportion of variance attributable to common DNA variants of allele frequency typically >1%. Hence $h^2_{SNP}$ is, by definition, lower than heritability gained from family and twin designs. The ratio of SNP-based to total heritability provides a description of the relative importance of common variants to the genetic architecture, as is vizualized in **figure 1B** 


```{r, h2_fam_SNP, echo=T, fig.width=5, fig.height=5, fig.align="center"}
heritability <- matrix(c(70,65,35,75,50,80,22.2,18.2,8.5,22.2,11.3,19.5),byrow = T, nrow = 2) #see table for refs
colnames(heritability) <- c("SCZ","BIP", "MDD","ADHD","ASD", "AN" )
rownames(heritability) <- c("h2_family", "h2_SNP")

l3<- c(as.expression((bquote(italic(h)[family]^2))),
       as.expression((bquote(italic(h)[SNP]^2))))

barplot(as.matrix(heritability),mgp=c(2,1,0), main="", beside=T,
        xlab="", ylab = expression(Heritability~italic((h^2))~'in'~'%'),ylim = c(0,100),
        col=c( "#8DA2B6", "#4E78A0"),  legend.text = l3, args.legend = list(bty = "n", x = "topright", ncol = 1, cex=0.8))
```


SNP-based heritability is estimated from genome-wide association study (GWAS) data, either directly from individual level genotype data or from GWAS summary statistics 

## GREML

To estimate the proportion of phenotypic variance captured by genotyped SNPs from individual-level data, Genome-based Restricted Maximum-likelihood analysis (GREML) is frequently used:

$$
y = covariates + g+e, with~V(y)= A_g\sigma^2_g + I\sigma^2_e
$$

Where $y$ is the phenotype, covariates could include age, sex and ancestry principal components (PCs), $g$ is the aggregate effect of all genome-wide SNPs for an individual, and $e$ is the residual effects for an individual. $V~(y)$ is the variance of $y$,  $\sigma^2_g$ is the variance of the g values, with the matrix $A_g$ being the genetic relationship matrix (GRM) since $g$ values are correlated between people, and $I$ is the identity matrix, since residual effects are assumed to be uncorrelated between individuals. The $h^2_{SNP}$ is then estimated as $h^2_{SNP} = \frac{\sigma^2_g}{\sigma^2_g + \sigma^2_e}$.

## GRM matrix

A key component of the GREML approach is the GRM ($A_g$) matrix, so lets have a closer look to it. 

Humans have diploid genomes, so at each biallelic SNP individuals can be homozygous for the reference allele, heterozysgous or homozygous for the non-reference (or alternative) allele, and genotypes for each SNP for an individual can be coded as 0,1,2 with respect to the count of alternative alleles. Under a random mating assumption, the number of alternative alleles for a SNP and individual follows a binomial distribution with 2 draws (one from the father and one from the mother) and probability equal to the minor allele frequency of that SNP.

Let’s assume that the minor allele frequency (MAF) of our $M$ SNPs come from a uniform distribution between 0 and 0.5.
We can then generate genotypes for one person using:

```{r, echo =T}
set.seed(666)
m <- 1000                                 # number of SNPs
maf <- runif(m, 0, .5)                    # random MAF for each SNP
x012 <- rbinom(m, 2, maf)
n <- 500                                    # number of individuals
X012 <- t(replicate(n, rbinom(m, 2, maf)))  # n x m genotype matrix (500 individuals X 1000 SNPs)
```

Monomorphic SNPs (SNPs of which frequencies do not vary between individuals in the sample) can cause problems in GRM analyses and should be removed. To end up with sufficient polymorphic SNPs, more SNPs will be generated and  only $M$ polymorphic SNPs will be remained (SNPs that vary in frequency between individuals)

```{r}
X012 = t(replicate(n, rbinom(2*m, 2, c(maf, maf))))
polymorphic = apply(X012, 2, var) > 0
X012 = X012[,polymorphic][,1:m]
maf = c(maf, maf)[polymorphic][1:m]
round(maf[1:10], 2)
```

```{r}
X012[1:5, 1:10]
```

A GRM make use of a scaled genotype matrix in which the genotypes of each SNP in our sample have mean 0 and variance 1, therefore the genotype matrix created above has to be scaled

```{r}
X = scale(X012, scale=TRUE)    
```

With the scaled genotype matrix, a GRM can be calculated using:

$$
A = \frac{XX^t}{M}
$$

```{r}
grm = (X %*% t(X))/m
round(grm[1:10,1:10],3)
```

Here, the offdiagonal elements represent the relatedness between two indivuals, while the diagonal elements represents the relatedness of an individual with itself, which is the average homozygosity or the level of inbreeding. If we visualize this:

```{r}
layout(matrix(c(1,2),1,2))
hist(grm[upper.tri(grm)],  ylab = "count", xlab = "value", col ="darkblue", main = "relatedness")
hist(diag(grm), ylab = "count", xlab = "value", col = "darkblue", main = "average homozygosity")
```

We see that the relatedness is centered around 0 (left panel) and the relatedness with itself is centered around 1 (right panel, where y-axis is cut-off at 300 for zoom-in effect). 

## LD Score regression

LD score regression (LDSC) was the first of now many methods to estimate SNP-based heritability from GWAS summary statistics. LDSC assumes a polygenic model of genetic effects on a trait. Under this assumption and recognising that there is local correlation between DNA variants along chromosomes (a reflection of past population growth and selection), a SNP that is correlated to other DNA variants (has a high linkage disequilibrium (LD) score) and is expected to have a higher association test statistic, on average. 

Theory shows that the expected ($E[~]$) relationship between the chi-square test statistics $(\chi^2)$) of SNPs and the LD score of SNPs is a function of the SNP-based heritability ($h^2_{SNP}$), sample size ($N$) and the number of SNPs ($M$) 

$$
E[\chi^2] = \frac{Nh^2_{SNP}}{M}l_j + intercept
$$

Using GWAS summary statistics, $h^2_{SNP}$ can be estimated from the regression coefficient (e.g. $\frac{Nh^2_{SNP}}{M}$) when the observed GWAS $\chi^2$ test statistics are regressed on the LD scores ($l_j$).

## Linkage disequilibrium (LD) Matrix

SNPs are often correlated with one another. Especially when they are physically close, since recombination rarely breaks up any correlation between them. This correlation between a pair of SNPs (two columns in our genotype matrix created earlier) is called linkage disequilibrium (LD) and can be estimated by calculating the correlation coefficient between the SNP genotypes.

The LD matrix contains the correlations of all SNP pairs in the genotype matrix and has therefore dimensions M×M. 

$$
cor(x,y) = \frac{cov(x,y)}{\sqrt{var(x)var(y)}}
$$

Since correlation implicitly scales by the variance it actually does not matter whether the scaled (X) or unscaled (X_012) genotype matrix is used. Further, when $X$ is already scaled, the covariance matrix is the same as the correlation matrix, the LD matrix can be calculated as:

$$
L = \frac{X^tX}{N}
$$

Therefore, if we calculate the LD matrices from our scaled or unscaled genotype matrices, we end up with approximately similar results:

```{r}
ld1 = cor(X012)
ld2 = cor(X)
mean(ld1)
mean(ld2)
```

These are close to zero, because we simulated the SNPs to be independent

## LD Scores

LD scores. $l_j$ are defined as the sum of squared correlations of a SNP $j$ with al other SNPs within a predefined region (usually, 1-Cm):

$$
lj = \sum_{k= 1}^{M} cor^2(X_j,X_k)
$$

As $X$ is standardized, we can calculate the sum of the squared sample correlations:

$$
\tilde{l_j} =\frac{1}{N^2}X^t_jXX^tX_j 
$$


To correct for bias (just as in the LDSC formula) we can:

$$
l_j = \frac{\tilde{l_j}N -M}{N + 1}
$$

```{r,fig.width=7, fig.height=4, fig.align="center"}
ldscores_sample = colSums(ld1^2)
ldscores = (ldscores_sample*n - m) / (n + 1)
head(ldscores_sample)
head(ldscores)
layout(matrix(c(1,2),1,2))
hist(ldscores, col = "darkblue", xlab = "Corrected scores", main = "")
hist(ldscores_sample, col = "lightgrey", xlab = "Uncorrected scores", main = "")
```

# Cross-disorder risk in relatives 

Just as epidemiological studies can investigate the increased risk of disorder A in relatives of those with disorder A, they also collect the data to estimate the increased risk of disorder B in relatives of those with disorder A.

Calculating the cross-disorder risk in relatives can be done in a rather similar way as described above using theory described in [Tallis 1987](https://www.ncbi.nlm.nih.gov/pubmed/24241289). Here, the threshold $T_{Rx,y}$ that defines $K_{Rx,y}$ as the lifetime risk of disease $y$ in relatives of probands with disease $x$ is defined as:
$$
T_{Rx,y} = \frac{T_x - a_{R}\sigma_{x,y}i_{y}}{\sqrt{1-a_{R}^2\sigma^2_{x,y}i_y(i_{y}-T_{y})}},
$$

where $r_{g}$ is the genetic correlation between the traits, and $r_{g}h_{x}h_{y}$ is the co-heritability between the traits, denoted $h_{x,y}$, and where and $i_x$ and $i_y$ , are the mean phenotypic liabilities of the two diseases. If the heritabilities of two diseases, their lifetime risks and the genetic correlation between them is known, then this equation can be used to estimate $K_{Rx,y}$ as:

$$
K_{Rx,y} = \Phi^{-1}(T_{Rx,y})
$$
where $\Phi^{-1}(x)$ is the inverse of the cummulative standard normal distribution function. And the risk ratio for relatives is defined as: 

$$
RR_{Rx,y} = \frac{K_{Rx,y}}{K_x}
$$

In R, you can calculate the cross disorder life-time risk and risk ratio of disease $x$ in relatives of probands with disease $y$ using the code below

```{r function get_CDRR}
h2x <- 0.7
h2y <- 0.35
Kx <- 0.01
Ky <- 0.15
rg <- 0.34
a <- 0.5
disorder1 <- "schizophrenia"
disorder2 <- "major depressive disorder"

get_CDRR = function(h2x, h2y, Kx, Ky, rg, a, disorder1, disorder2){
  cat("Input Parameters:\n")
  cat("----------------------------\n")
  cat("disorder1:", disorder1, "\n")
  cat("h2x: ",h2x,"\n")
  cat("Kx: ",Kx,"\n")
  cat("disorder2:", disorder2, "\n")
  cat("h2y: ",h2y,"\n")
  cat("Ky: ",Ky,"\n")
  cat("IBD: ",a, "\n")
  cat("----------------------------\n")    

  Tx = -qnorm(Kx, 0, 1) 
  Ty = -qnorm(Ky, 0, 1) 
  zx = dnorm(Tx) 
  zy = dnorm(Ty) 
  ix = zx / Kx 
  iy = zy / Ky
  
  # calculate the genetic covariance:
  covg = rg * sqrt(h2x * h2y)
  
  # calculate risk ratio for relatives
  Txy  = (Tx - a * covg * iy) / sqrt(1 - a^2 * covg^2 * iy * (iy-Ty))
  Kxy  = 1 - pnorm(Txy)
  RRxy = Kxy/Kx # relative risk of disease x given that parent has disease y


cat("Output Parameters:\n")
cat("----------------------------\n")
cat("Kxy: ",round(Kxy,2), "\n")
cat("RRxy: ",round(RRxy,2), "\n")
cat("----------------------------\n")  
}

get_CDRR(h2x, h2y, Kx, Ky, rg, a, disorder1, disorder2)

```


# Genetic correlation
If from real data the lifetime risks of the two disorders are known, i.e., $K_{x}$ and $K_{y}$ then these can be used to calculate the thresholds $T_x$, $T_y$. If the risks in relatives are measured of disease x in relatives of disease x, $K_{Rx,x}$ and of disease y in relatives of disease y, $K_{Rx,x}$ these can be used to calculate the thresholds of  $T_{Rx,x}$ and $T_{Ry,y}$, then these can be used to estainte the heritability of each disorder. Last if the lifetime risk of disease x is measured in relatives of those with disease y $K_{Rx,y}$ then this can be used to calculate and $T_{Rx,y}$ then these can be used to calculate the genetic correlation between the two disorders, by making $r_g$ the subject of the equation above


$$
r_g = \frac{T_y-T_{Ry,x}\sqrt{1-(1-T_x/i_x)(T_{y}^2-T_{Ry,x}^2)}}{a_R (i_x+(i_x-T_x )T_{Ry,x}^2)\sqrt{h^2_x h^2_y}}
$$

# SNP-based genetic correlation
Bivariate extensions of both the GREML and LDSC methods allow estimation of the a SNP-based genetic correlation from GWAS data sets that have been collected independently for the two traits. 

## bivariate GREML
In essence, bivariate GREML detects if cases of the two diseases are significantly more similar genetically than they are to controls (or significantly less similar in the case of negative correlation) ([Lee et al 2012](Hh3ecY_mOD407DOoMHAdYzn69dBQzX1fdElmLzY4jlU4TthxoyHbNe_898Ur1vCMQZG7FQx1epY5ADDKhupCUBSVM9SrvHsWj_sFVS7ytVxTCEXR3VJdzusRMdl3Tj55d1NmUPErz74jwyPfutOrqkPhj4g3tlri9jHEgsB0SoPUMMjYwZOgs86dQb5Q))

Consider $y_1 = X_1b_1 + Z_1g_1 +e_1$ for trait 1 and $y_2 = X_2b_2 + Z_2g_2 +e_2$ for trait 2.

Here, $y_1$ and $y_2$) are two vectors of observations for trait 1 and 2, $b_1$ and $b_2$ are vectors of fixed effects, $g_1$ and $g_2$ are vectors of random polygenic effects for each individual in both trait 1 and trait2, while $e_1$ and $e_2$ the residuals for trait 1 and trait 2 are. Additionally, $X$ and $Z$ are incidence matrices for the effects $b$ and $g$. The variance covatiance matrix ($V$) is subsequently defined as:

$$
V = 
\begin{pmatrix}
Z_1AZ'_1\sigma^2_{g1} + I\sigma^2_{e1}&Z_1AZ'_2\sigma^2_{g1g2}\\
Z_2AZ'_1\sigma^2_{g1g2}&Z_2AZ'_2\sigma^2_{g2} + I\sigma^2_{e2}
\end{pmatrix}
$$
 
Where $A$ is the genomic relatedness matrix [see above](#grm-matrix) and $I$ is an identity matrix, $\sigma^2_{g},\sigma^2_e$ and $\sigma^2_{g1g1}$, which are respectively the genetic variance, residual variance and covaraince ebtween $g1$ and $g2$.

## Case Control Disease status

For disease traits when $y$ is a vector consisting only from 1s (cases) and 0s (controls), the [liability threshold model](#liability-threshold-model) can be written to transform the observed discrete scale to an unobserved continuous liability using [Falconer, 1965](https://onlinelibrary-wiley-com.vu-nl.idm.oclc.org/doi/pdf/10.1111/j.1469-1809.1965.tb00500.x), [Dempster and Lerner 1950](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1209482/pdf/212.pdf)

$$
l = g^* + e^*
$$

where $l$ is a vector of liability phenotypes wich are normally distributed ($N(0,1)$) in the population. Subsequently, $e^*$ is a vector of random residuals on the liability scale distributed with $N(0,\sigma^2_{e*})$. Using this linear appoximation, the correlation between a pair of disorders is similar on the observed and liability scale. As demonstrated by [Lee et al. (2010)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3059431/pdf/main.pdf), the genetic value on the observed scale can be corrected when samples are ascertained (which is often the case in case-control studies) using:

$$
g_{cc}\cong c+z \frac{P(1-P)}{K(1-K)}g^*
$$

where $gcc$ is genetic values on the observed scale in a case-control study, $c$ is a constant, $K$ is the lifetime risk  in the general population, and $P$ is the proportion of the sample that are cases while $z$ is the height of the standard normal probability desnity function that indicates the proportion $K$ (height of the normal distribution curve at the threshold boundary). From this, the covariance between genetic values of two disorder traits with ascertained samples can be estimated using:

$$
cov(g_{cc1},g_{cc2})\cong z_1 \frac{P_1(1-P_1)}{K_1(1-K_1)}z_2\frac{P_2(1-P_2)}{K_2(1-K_2)}cov(g^*_1,g^*_2)
$$

From this information it becomes clear that even when samples are ascertained, the correlation is the same as unbiased samples due to the approximate linear relatioinship between the genetic values on the different scales.

## Genetic correlation using LD score regression
LD Score regression [Bulik-Sullivan et al. 2015](https://www-nature-com.vu-nl.idm.oclc.org/articles/ng.3406.pdf) on the other hand, estimates the genetic correlation from GWAS sumary statistics. The method is based on the knowlegde that the GWAS effect size estimate for a given SNP incorporates the effect of all SNPs in linkage disequilibrium with that SNP [click here for more information on LD scores](#ld-score-regression). For complex traits, SNPs with high LD-scores have on average higher $\chi^2$ statistics compared to SNPs with low LD-scores under a polygenic model. This observation holds if the $\chi^2$ statistic representing a single trait is replaced with the product of the $z$ scores from two traits (bivariate genetic correlation). Here the expected ($E$) value of $z_{1j}z_{2j}$ for $SNP_j$ is:

$$
E[z_{1j}z_{2j}l_j] = \frac{\sqrt{N_1N_2h_{1,2}}}{M}l_j + \frac{\varrho N_s}{\sqrt{N_1N_2}}
$$

Here, $N_i$ is the sample size for the included studies, $h_{1,2}$ is the genetic covariance and $l_j$ is the LD score. $N_s$ represents the sample overlap between both studies and $\varrho$ is the phenotypic correlation between the overlapping samples. 

In general, sample overlap generates a correlation between $z_{1j}$ and $z_{2j}$, which inflates $z_{1j}z_{2j}$. However, the effect of this inflation is uniform accorss all markers in is independent of the LD scores. Therefore, sample overlap only affects the intercept, which is represented by $\varrho N_s/\sqrt{N_1N_2}$. The slope itself will not be biased by sample overlap 

# Simulations

(Simulations phenotypes partly based on [van Rheenen et al., 2019](https://www-nature-com.vu-nl.idm.oclc.org/articles/s41576-019-0137-z)

Simulation can be a useful tool. Here we simulate phenotypes (disease liability and case/control status) in two correlated disorders SCZ ($h^2_x$)  and MDD ($h^2_y$) in $N$ parent-offspring pairs ($a_R$ = 0.5) Phenotypic liability is simulated as $P = G + E$, where $G$ is the genetic value and $E$ the non-genetic (residual) value. $G$ is drawn from the multivariate normal distribution $N($**0**$, V_g)$ where  ($V_g$) is a symmetric variance-covariance matrix:
$$
V_g = 
\begin{pmatrix}
h^2_{x}&\rho_{gx,y}\sqrt{h^2_{x}h^2_{y}}&\frac{1}{2}h^2_{x}& \frac{1}{2}\rho_{gx,y}\sqrt{h^2_{x}h^2_{y}}\\
&h^2_{y}&\frac{1}{2}\rho_{gx,y}\sqrt{h^2_{x}h^2_{y}}&\frac{1}{2}h^2_{y} \\
& & h^2_x &\rho_{gx,y}\sqrt{h^2_{x}h^2_{y}} & \\
& & & h^2_y
\end{pmatrix}
$$

Where $h^2$ is there heritability of the trait and $\rho_g$ is the genetic correlation. Note, that shared non-genetic effects are not modelled, such that $E$ are drawn from multivariate lnormal distribution $N($**0**$, V_e)$ where ($V_e$) is a symmetric variance-covariance matrix:
$$
V_e = 
\begin{pmatrix}
(1-h^2_{x})& 0 & 0 & 0 \\
&(1-h^2_{y})& 0 & 0 \\
& & (1-h^2_x) & 0 \\
& & & (1-h^2_y)
\end{pmatrix}
$$

```{r heritability and rg estimates}
N=1000000
Kx=0.01 
Ky=0.15 
h2x=0.7 
h2y=0.35 
rg=0.34
a =0.5

h2_estimator = function(N, Kxp, Kyp, h2x, h2y, rg,a){

  cat("Input Parameters:\n")
  cat("----------------------------\n")
  cat("h2x  : ", h2x, "\n")
  cat("h2y  : ", h2y, "\n")
  cat("rg   : ", rg, "\n")
  cat("IBD  : ", a, "\n")
  cat("N    : ", N, "\n")
  cat("Estimates:\n")
  cat("----------------------------\n")
  
  # simulate the phenotypes for x and y for parent (p) and child (c)
  # define genetic parameters
  gcovxy = rg*sqrt(h2x*h2y)

  Vg  = matrix(c(h2x, gcovxy, a*h2x, a*gcovxy,
                 gcovxy, h2y, a*gcovxy, a*h2y,
                 a*h2x,  a*gcovxy, h2x, gcovxy,
                 a*gcovxy, a*h2y, gcovxy, h2y), nrow=4)
  
  # simulate genetic values
  G  = MASS::mvrnorm(n=N, mu=c(0,0,0,0), Sigma=Vg)
  Vg  = matrix(c(h2x, gcovxy, a*h2x, a*gcovxy,
                 gcovxy, h2y, a*gcovxy, a*h2y,
                 a*h2x,  a*gcovxy, h2x, gcovxy,
                 a*gcovxy, a*h2y, gcovxy, h2y), nrow=4)
  
  # simulate genetic values
  G  = MASS::mvrnorm(n=N, mu=c(0,0,0,0), Sigma=Vg)
  
  # simulate trait liability for x and y
  Yxp = G[,1] + rnorm(N, mean=0, sd=sqrt(1-h2x))
  Yyp = G[,2] + rnorm(N, mean=0, sd=sqrt(1-h2y))
  Yxc = G[,3] + rnorm(N, mean=0, sd=sqrt(1-h2x))
  Yyc = G[,4] + rnorm(N, mean=0, sd=sqrt(1-h2y))

  # define parameter from normal distribution theory
  Tx = -qnorm(Kx, 0, 1)
  Ty = -qnorm(Ky, 0, 1)
  zy = dnorm(Ty)
  zx = dnorm(Tx)
  iy = zy/Ky
  ix = zx/Kx

  # dichotomize to define binary trait
  Yxp_cc = rep(0, N) ; Yxp_cc[Yxp >= Tx] = 1
  Yyp_cc = rep(0, N) ; Yyp_cc[Yyp >= Ty] = 1
  Yxc_cc = rep(0, N) ; Yxc_cc[Yxc >= Tx] = 1
  Yyc_cc = rep(0, N) ; Yyc_cc[Yyc >= Ty] = 1

  # estimate heritability using normal distribution theory:
  Kx_est = sum(Yxp_cc, Yxc_cc) / (2*N)
  Ky_est = sum(Yyp_cc, Yyc_cc) / (2*N)

  Tx_est = -qnorm(Kx_est, 0, 1)
  zx_est = dnorm(Tx_est)
  ix_est = zx_est / Kx_est

  Ty_est = -qnorm(Ky_est, 0, 1)
  zy_est = dnorm(Ty_est)
  iy_est = zy_est / Ky_est

  Trxx_est  = -qnorm(sum(Yxc_cc[Yxp_cc == 1]) / sum(Yxp_cc), 0, 1)
  Tryy_est  = -qnorm(sum(Yyc_cc[Yyp_cc == 1]) / sum(Yyp_cc), 0, 1)

  h2x_est = (Tx_est - Trxx_est * sqrt(1-(1 - Tx_est/ix_est)*(Tx_est^2 - Trxx_est^2)) ) / (a * (ix_est + (ix_est - Tx_est) *     Trxx_est^2))
  
  h2y_est = (Ty_est - Tryy_est * sqrt(1-(1 - Ty_est/iy_est)*(Ty_est^2 - Tryy_est^2)) ) / (a * (iy_est + (iy_est - Ty_est) *     Tryy_est^2))
  
  # estimate the genetic correlation using normal distribution theory:
  Tryx_est = -qnorm(sum(Yyc_cc[Yxp_cc == 1]) / sum(Yxp_cc), 0, 1)
  rg_est = (Ty_est - Tryx_est * sqrt(1 - (1 - Tx_est/ix_est)*(Ty_est^2 - Tryx_est^2)))  / (a*(ix_est + (ix_est - Tx_est) *Tryx_est^2) *      sqrt(h2x_est * h2y_est))

cat("Output Parameters:\n")
cat("----------------------------\n")
cat("Kx_est : ", round(Kx_est,2), "\n")
cat("Ky_est : ", round(Ky_est,2), "\n")
cat("h2x_est: ",round(h2x_est,2),"\n")
cat("h2y_est: ",round(h2y_est,2),"\n")
cat("rg_est : ",round(rg_est,2),"\n")
cat("----------------------------\n")  
}

h2_estimator(N, Kxp, Kyp, h2x, h2y, rg,a)
```

Note, if you increase $N$ ($N$ was kept relatively low for computational reasons), the estimated values ($h^2_x$, $h^2_y$, and $rg$) wil come closer to the input data.
